File: CPP/GameOfLife/Cell.cpp
#include <cstdio>

#include "Cell.h"

namespace impl {
	Cell::Cell() {
		this->is_alive = false;
		this->is_alive_next = false;
		this->changed = false;
	}

	// ===========================================================================================================
	void Cell::Step() {
		is_alive = is_alive_next;
	}

	// ===========================================================================================================
	void Cell::CalculateNextStep(const unsigned short alive_neighbours) {
		if(is_alive)
			is_alive_next = alive_neighbours == 2 || alive_neighbours == 3;
		else
			is_alive_next = alive_neighbours == 3;

		if(is_alive != is_alive_next)
			changed = true;
	}

	// ===========================================================================================================
	void Cell::Draw() const {
		// TODO: Maybe print emojis (if I get it working) - Unicode not detected atm.
		printf(is_alive ? "■" : "□");
	}

	// ===========================================================================================================
	bool Cell::IsAlive() const {
		return is_alive;
	}

	// ===========================================================================================================
	bool Cell::Changed() const {
		return changed;
	}

	// ===========================================================================================================
	void Cell::SetAlive(const bool alive) {
		is_alive = alive;
	}
}

File: Go/cell/cell.go
package cell

import "fmt"

// Starting with capital letter --> Public
// Starting with lowercase letter --> Private

type Cell struct {

	// Public
	IsAlive    bool
	HasChanged bool

	// Private
	isAliveNext bool
}

// Think of it as a constructor but as a separate function
func New(isAlive bool) Cell {
	c := Cell{
		IsAlive:     isAlive,
		HasChanged:  true,
		isAliveNext: isAlive,
	}

	return c
}

// (c Cell) is the receiver, this means that you "call" this function on a Cell-type object
func (c *Cell) CalculateNextStep(aliveNeighbours byte) {
	if c.IsAlive {
		c.isAliveNext = aliveNeighbours == 2 || aliveNeighbours == 3
	} else {
		c.isAliveNext = aliveNeighbours == 3
	}

	if c.IsAlive != c.isAliveNext {
		c.HasChanged = true
	}
}

func (c *Cell) Step() {
	c.IsAlive = c.isAliveNext
}

func (c *Cell) Hello() {
	fmt.Println("Hello, world. I'm a cell. Am I currently alive? ", c.IsAlive)
}

func (c Cell) Draw() {
	if c.IsAlive {
		fmt.Print("■")
	} else {
		fmt.Print("□")
	}
}
